### PR TITLE
SAAS-4977: Added EBS Volumes Optimized Plugins

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

**Comment by sadeed12345** [comment]
Permissions Used For the Plugin: 
"compute-optimizer:GetRecommendationSummaries",
"compute-optimizer:GetEC2InstanceRecommendations",
"compute-optimizer:GetAutoScalingGroupRecommendations"

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+            if (getRecommendationSummaries.err || !getRecommendationSummaries.data) {
+                helpers.addResult(results, 3,
+                    'Unable to get recommendation summaries: ' + helpers.addError(getRecommendationSummaries), region);
+                return rcb();
+            }
+
+            if (!getRecommendationSummaries.data.length) {
+                helpers.addResult(results, 0, 
+                    'Optimization for summaries is not configured', region);
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.js:40]:
```suggestion
                    'No Compute Optimizer optimization summaries found', region);
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.spec.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            });
+        });
+
+        it('should PASS if Optimization for summaries is not configured', function (done) {
+            const cache = createCache([]);
+            ebsVolumesOptimized.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Optimization for summaries is not configured');
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.spec.js:122]:
```suggestion
                expect(results[0].message).to.include('No Compute Optimizer optimization summaries found');
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.spec.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            ebsVolumesOptimized.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('EBS volumes have no recommendations enabled');
+                done();
+            });
+        });
+
+        it('should PASS if Optimization for summaries is not configured', function (done) {
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.spec.js:116]:
```suggestion
        it('should PASS if No Compute Optimizer optimization summaries found', function (done) {
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                } else if (notOptimized.value){
+                    helpers.addResult(results, 2,
+                        `EBS volumes are not optimized,  NOT_OPTIMIZED: ${notOptimized.value}`, region);
+                } else {
+                    helpers.addResult(results, 0,
+                        'All EBS volumes are optimized', region);
+                }
+            } else {
+                helpers.addResult(results, 2,
+                    'No EBS volumes configured', region);
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.js:62]:
```suggestion
                    'Recommendation summaries are not configured for EBS Volumes', region);
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'EBS Volumes Optimized',
+    category: 'Compute Optimizer',
+    domain: 'Management and Governance',
+    description: 'Ensure that Compute Optimizer EBS volume recommendations are in order to take the actions to optimize Amazon EBS volumes that are under-performing.',
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.js:8]:
```suggestion
    description: 'Ensure that Compute Optimizer EBS volume recommendations are enabled.',
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+            let findings = getRecommendationSummaries.data.find(resourceType => resourceType.recommendationResourceType === 'EbsVolume');
+            if (findings) {
+      
+                let notOptimized = findings.summaries.find(notOpt => notOpt.name === 'NotOptimized');
+                let optimized = findings.summaries.find(opt => opt.name === 'Optimized');
+                
+                if (!notOptimized.value  && !optimized.value){
+                    helpers.addResult(results, 0,
+                        'EBS volumes have no recommendations enabled', region);
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.js:52]:
```suggestion
                        'No recommendations found for EBS volumes', region);
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.spec.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            });
+        });
+
+        it('should PASS if EBS volumes have no recommendations enabled', function (done) {
+            const cache = createCache([getRecommendationSummaries[2]]);
+            ebsVolumesOptimized.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('EBS volumes have no recommendations enabled');
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.spec.js:111]:
```suggestion
                expect(results[0].message).to.include('No recommendations found for EBS volumes');
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.spec.js [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            ebsVolumesOptimized.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('EBS volumes are not optimized');
+                done();
+            });
+        });
+
+        it('should PASS if EBS volumes have no recommendations enabled', function (done) {
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.spec.js:105]:
```suggestion
        it('should PASS if EBS volumes have no recommendations found for EBS volumes', function (done) {
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+      
+                let notOptimized = findings.summaries.find(notOpt => notOpt.name === 'NotOptimized');
+                let optimized = findings.summaries.find(opt => opt.name === 'Optimized');
+                
+                if (!notOptimized.value  && !optimized.value){
+                    helpers.addResult(results, 0,
+                        'No recommendations found for EBS volumes', region);
+                } else if (notOptimized.value){
+                    helpers.addResult(results, 2,
+                        `EBS volumes are not optimized,  NOT_OPTIMIZED: ${notOptimized.value}`, region);
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.js:55]:
```suggestion
                        `Found ${notOptimized.value} unoptimized EBS volumes`, region);
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.spec.js [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            });
+        });
+
+        it('should FAIL if EBS volumes are not optimized', function (done) {
+            const cache = createCache([getRecommendationSummaries[1]]);
+            ebsVolumesOptimized.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('EBS volumes are not optimized');
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.spec.js:100]:
```suggestion
                expect(results[0].message).to.include('unoptimized EBS volumes');
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'EBS Volumes Optimized',
+    category: 'Compute Optimizer',
+    domain: 'Management and Governance',
+    description: 'Ensure that Compute Optimizer EBS volume recommendations are enabled.',
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.js:8]:
```suggestion
    description: 'Ensure that Compute Optimizer recommendation summaries does not include unoptimized EBS Volumes.',
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'EBS Volumes Optimized',
+    category: 'Compute Optimizer',
+    domain: 'Management and Governance',
+    description: 'Ensure that Compute Optimizer recommendation summaries does not include unoptimized EBS Volumes.',
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.js:8]:
```suggestion
    description: 'Ensure that Compute Optimizer does not have active recommendation summaries for unoptimized EBS Volumes.',
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'EBS Volumes Optimized',
+    category: 'Compute Optimizer',
+    domain: 'Management and Governance',
+    description: 'Ensure that Compute Optimizer does not have active recommendation summaries for unoptimized EBS Volumes.',
+    more_info: 'An EBS volume is considered optimized when Compute Optimizer determines that the volume is correctly provisioned to run your workload, based on the chosen volume type, volume size, and IOPS specification. For optimized resources, Compute Optimizer might sometimes recommend a new generation volume type.',
+    link: 'https://docs.aws.amazon.com/compute-optimizer/latest/ug/view-ebs-recommendations.html',
+    recommended_action: 'Enable Compute Optimizer Opt In options for EBS volume recommendations',
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.js:11]:
```suggestion
    recommended_action: 'Resolve Compute Optimizer recommendations for EBS Volumes.',
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'EBS Volumes Optimized',
+    category: 'Compute Optimizer',
+    domain: 'Management and Governance',
+    description: 'Ensure that Compute Optimizer does not have active recommendation summaries for unoptimized EBS Volumes.',
+    more_info: 'An EBS volume is considered optimized when Compute Optimizer determines that the volume is correctly provisioned to run your workload, based on the chosen volume type, volume size, and IOPS specification. For optimized resources, Compute Optimizer might sometimes recommend a new generation volume type.',
+    link: 'https://docs.aws.amazon.com/compute-optimizer/latest/ug/view-ebs-recommendations.html',
+    recommended_action: 'Resolve Compute Optimizer recommendations for EBS Volumes.',
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.js:11]:
```suggestion
    recommended_action: 'Resolve Compute Optimizer recommendations for EBS volumes.',
```

File: plugins/aws/computeoptimizer/ebsVolumesOptimized.js [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        var regions = helpers.regions(settings);
+
+        async.each(regions.computeoptimizer, function(region, rcb){
+            var getRecommendationSummaries = helpers.addSource(cache, source,
+                ['computeoptimizer', 'getRecommendationSummaries', region]);
+
+            if (!getRecommendationSummaries) return rcb();
+
+            if (getRecommendationSummaries && getRecommendationSummaries.err && 
+                getRecommendationSummaries.err.code === 'OptInRequiredException'){
```
Comment (AkhtarAmir) [inline plugins/aws/computeoptimizer/ebsVolumesOptimized.js:26]:
```suggestion
                getRecommendationSummaries.err.code.toUpperCase() === 'OPTINREQUIREDEXCEPTION'){
```

**Pull request merged**